### DISCUSSION TITLE
What is preventing topological sorting of a multigraph?

### DISCUSSION THREAD

**Discussion by TaylorBenner** [graphology/graphology#498] [Q&A] [answered]
I'm rather new to graph theory and this library.  I'm hoping to find a way to topologically sort a multigraph with parallel edges, and I'm wondering why this is expressly prohibited in the topological sort implementation in the dag package?  I would appreciate any eli5 explanations people could offer, so I could go about solving this in my own code.

**Comment by Yomguithereal** [comment]
I don't think there is any theoretical reasons why one should not be able to topologically sort a multi graph. It's just that in this precise case, sorting a multi graph or a simple projection of a multi graph (where you only keep one instance of edge between a pair of nodes) is essentially equivalent. So it's probably a performance or ease of implementation thing, because iterating over unique neighbors and computing the "simple" degree of a node in a multi graph has some hidden linear cost. What's more, nobody spoke of this use case (wrt graphology) before you did so there was no incentive to think about the multi graph implementation :)

It is therefore probably doable to adapt the current implementation so it can support multigraphs, but it must be done cleverly and it may complexify the initial degree accumulation a little bit.

A workaround, currently, would just be to use the `toSimple` operator from here to pass a simple projection to the topological sorting functions: https://graphology.github.io/standard-library/operators.html#tosimple

In any case, can you elaborate a little bit more on your use case, I am curious.

**Reply by TaylorBenner** [reply]
Thanks for your quick, and detailed response.  My use case is to support a data flow programming model using the graphology library to handle the more complex side of the DAG interactions.  This is commonly seen in programs like Blender or Unity for things like shader graphs.  

I'm looking to topologically sort a data flow graph in order to determine the input dependencies of any particular node in the DAG, so a stream can be created that performs the various transforms needed.

**Reply by Yomguithereal** [reply]
Would having the topological sorting functions support multigraphs useful to you then? Are you using multiple edges to represent complex interactions, or are they used only as an aggregation proxy?

**Reply by TaylorBenner** [reply]
The way my application is structured at the moment, a `toSimple` call may be what I need.  I suppose my problem is a simple dependency graph where upstream input is aggregated in order to execute a target node.  In that case, to your question, I suspect preserving multiple edges during a topological sort is unnecessary at the moment.

**Reply by Yomguithereal** [reply]
What I mean is that maybe in your case you could keep a simple graph and increment/aggregate some value stored as an attribute of your edges instead of using a multi graph. But it is also possible that it is not an option or not convenient in your case, typically if your edges have discrete metadata such as a type or whatnot.

**Comment by Yomguithereal** [comment, marked as answer]
@TaylorBenner v0.4.0 of `graphology-dag` with multi graph support was released just now.
